### PR TITLE
qemu_mem_utilization:qemu mem utilization check during guest reboot

### PR DESCRIPTION
**Pull request by aliang123**
qemu_mem_utilization:qemu mem utilization check during guest reboot

tests.qemu_mem_utilization:
1.add function for qemu memory check during reboot/shutdwon with blk

test.qemu_mem_utilization:
1.define cfg for reboot and shutdown test for times.

Signed-off by:Aihua Liang<aliang@redhat.com>

ID:1134275

### CODE REVIEW

File: qemu/tests/cfg/qemu_mem_utilization.cfg [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+    repeat_times = 1000
+    kill_vm = yes
+    kill_vm_gracefully = no
+    host_mem = "ps aux|grep qemu|grep -v grep|awk '{print $4}'|sed -n '3p'"
+    variants:
+        - with_system_reset:
+            sub_test = boot
+            reboot_method = system_reset
+            sleep_before_reset = 20
+        - with_shutdown:
```
Comment (luckyh) [inline qemu/tests/cfg/qemu_mem_utilization.cfg]: IMHO, the shutdown test isn't meaningful.

File: qemu/tests/qemu_mem_utilization.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    :param env: Dictionary with test environment.
+    """
+
+    logging.info("Wait 90s for guest's up")
+    time.sleep(90)
+    error_context.context("Check $qemu memory utility before test", logging.info)
+    cmd = params.get("host_mem")
+    qemu_mem_utility_before = float(utils.system_output(cmd))
+
+    repeat_times = int(params.get("repeat_times", 10000))
```
Comment (CongLi) [inline qemu/tests/qemu_mem_utilization.py]: 1000 is fine, 10000 is too many.

File: qemu/tests/qemu_mem_utilization.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+            vm = env.get_vm(params["main_vm"])
+            timeout = float(params.get("login_timeout", 360))
+            session = vm.wait_for_login(timeout=timeout)
+        repeat_times -= 1
+    logging.info("Wait 90s for guest's up")
+    time.sleep(90)
+    error_context.context("Check $qemu memory utility after test", logging.info)
+    qemu_mem_utility_after = float(utils.system_output(cmd))
+    error_context.context("Compare $qemu's mem utility", logging.info)
+    if abs(qemu_mem_utility_after - qemu_mem_utility_before) > 1:
```
Comment (CongLi) [inline qemu/tests/qemu_mem_utilization.py]: it's strange, mem_diff > 1? 
could you please help explain it?

File: qemu/tests/qemu_mem_utilization.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    cmd = params.get("host_mem")
+    qemu_mem_utility_before = float(utils.system_output(cmd))
+
+    repeat_times = int(params.get("repeat_times", 10000))
+    sub_test = params.get("sub_test")
+    vm = env.get_vm(params["main_vm"])
+    error_context.context("Do subtest:%s repeatly" % sub_test, logging.info)
+    while repeat_times:
+        utils_test.run_virt_sub_test(test, params, env, sub_test)
+        if not vm.is_alive():
```
Comment (CongLi) [inline qemu/tests/qemu_mem_utilization.py]: Better to check crash here.

File: qemu/tests/qemu_mem_utilization.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    4) check $qemu memory utility
+
+    :param test: QEMU test object
+    :param params: Dictionary with the test parameters
+    :param env: Dictionary with test environment.
+    """
+
+    logging.info("Wait 90s for guest's up")
+    time.sleep(90)
+    error_context.context("Check $qemu memory utility before test", logging.info)
```
Comment (nasastry) [inline qemu/tests/qemu_mem_utilization.py]: $qemu?

Comment (aliang123) [inline qemu/tests/qemu_mem_utilization.py]: Typo error, should be "qemu", will update it, thanks.

File: qemu/tests/cfg/qemu_mem_utilization.cfg (hunk 1 above)
Comment (aliang123) [inline qemu/tests/cfg/qemu_mem_utilization.cfg]: Thanks for your advice, will update it later.

File: qemu/tests/qemu_mem_utilization.py (hunk 2 above)
Comment (aliang123) [inline qemu/tests/qemu_mem_utilization.py]: OK, will update it.

**Comment by aliang123** [comment]
Test cmds:
   python ConfigTest.py  --guestname=RHEL.7.4 --platform=x86_64 --verbose=yes --testcase=qemu_mem_utilization
Test Result:
   PASS 1-smp_2.512m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_blk.up.virtio_net.RHEL.7.4.x86_64.io-github-autotest-qemu.qemu_mem_utilization.with_system_reset
PASS 2-smp_2.512m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_blk.up.virtio_net.RHEL.7.4.x86_64.io-github-autotest-qemu.qemu_mem_utilization.with_reboot

**Comment by aliang123** [comment]
Test cmd:
   python ConfigTest.py --guestname=Win2016 --platform=x86_64 --verbose=yes --testcase=qemu_mem_utilization
Test Result:
    PASS 1-smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_blk.up.virtio_net.Win2016.x86_64.io-github-autotest-qemu.qemu_mem_utilization.with_system_reset
    PASS 2-smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_blk.up.virtio_net.Win2016.x86_64.io-github-autotest-qemu.qemu_mem_utilization.with_reboot

File: qemu/tests/qemu_mem_utilization.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    2) after guest up, check qemu memory utilization
+    3) reboot guest for times
+    4) check qemu memory utilization
+
+    :param test: QEMU test object
+    :param params: Dictionary with the test parameters
+    :param env: Dictionary with test environment.
+    """
+
+    def mem_utilization(timeout):
```
Comment (CongLi) [inline qemu/tests/qemu_mem_utilization.py]: doc string

File: qemu/tests/qemu_mem_utilization.py [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    error_context.context("Do subtest:%s repeatly" % sub_test, logging.info)
+    utils_test.run_virt_sub_test(test, params, env, sub_test)
+    if not vm.is_alive():
+        raise error.TestFail("Guest is dead during reboot test")
+
+    error_context.context("Check qemu memory utilization after test", logging.info)
+    qemu_mem_utilization_after = mem_utilization(timeout)
+
+    error_context.context("Compare qemu's mem utilization", logging.info)
+    if abs(qemu_mem_utilization_after - qemu_mem_utilization_before) > 1:
```
Comment (CongLi) [inline qemu/tests/qemu_mem_utilization.py]: 10%

File: qemu/tests/qemu_mem_utilization.py [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+    3) reboot guest for times
+    4) check qemu memory utilization
+
+    :param test: QEMU test object
+    :param params: Dictionary with the test parameters
+    :param env: Dictionary with test environment.
+    """
+
+    def mem_utilization(timeout):
+        cmd = params.get("host_mem")
```
Comment (CongLi) [inline qemu/tests/qemu_mem_utilization.py]: params[''host_mem]

File: qemu/tests/qemu_mem_utilization.py [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+        qemu_mem = float(utils.system_output(cmd))
+        end_time = time.time() + float(timeout)
+        while time.time() < end_time:
+            time.sleep(10)
```
Comment (CongLi) [inline qemu/tests/qemu_mem_utilization.py]: should move the time.sleep() out of the while loop, right?
sleep 10s every time?

File: qemu/tests/qemu_mem_utilization.py [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+    utils_test.run_virt_sub_test(test, params, env, sub_test)
+    if not vm.is_alive():
+        raise error.TestFail("Guest is dead during reboot test")
+
+    error_context.context("Check qemu memory utilization after test", logging.info)
+    qemu_mem_utilization_after = mem_utilization(timeout)
+
+    error_context.context("Compare qemu's mem utilization", logging.info)
+    if abs(qemu_mem_utilization_after - qemu_mem_utilization_before) > 1:
+        raise error.TestFail("Host Memory Leak exist, please check")
```
Comment (CongLi) [inline qemu/tests/qemu_mem_utilization.py]: Better to print the qemu mem utilization status in the error info

File: qemu/tests/qemu_mem_utilization.py [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+        else:
+            raise error.TestFail("Fail to get qemu_mem_utilization")
+
+    timeout = params.get("login_timeout", 120)
+    error_context.context("Check qemu memory utilization before test", logging.info)
+    qemu_mem_utilization_before = mem_utilization(timeout)
+
+    sub_test = params.get("sub_test")
+    vm = env.get_vm(params["main_vm"])
+    error_context.context("Do subtest:%s repeatly" % sub_test, logging.info)
```
Comment (CongLi) [inline qemu/tests/qemu_mem_utilization.py]: both 2 variants's sub test are boot, then no difference.
it's better to print the reboot method.

File: qemu/tests/qemu_mem_utilization.py (hunk 7 above)
Comment (aliang123) [inline qemu/tests/qemu_mem_utilization.py]: ok, i'll update it to 10 later.

File: qemu/tests/qemu_mem_utilization.py (hunk 8 above)
Comment (aliang123) [inline qemu/tests/qemu_mem_utilization.py]: The same usage, i think.

File: qemu/tests/qemu_mem_utilization.py (hunk 9 above)
Comment (aliang123) [inline qemu/tests/qemu_mem_utilization.py]: yes,  check in a 10-second interval, so will sleep 10s every time.

File: qemu/tests/qemu_mem_utilization.py (hunk 10 above)
Comment (aliang123) [inline qemu/tests/qemu_mem_utilization.py]: good idea, will update it later.

File: qemu/tests/qemu_mem_utilization.py (hunk 11 above)
Comment (aliang123) [inline qemu/tests/qemu_mem_utilization.py]: got it, will update it later.

File: qemu/tests/qemu_mem_utilization.py (hunk 8 above)
Comment (CongLi) [inline qemu/tests/qemu_mem_utilization.py]: No.
params['host_mem']: if 'host_mem' is empty, there will be error directly. it's always used for those must params.
Instead, there will no error for params.get('host_mem') although 'host_mem' is empty or not existed.

**Comment by aliang123** [comment]
Through discussion, the test method for the case is not acceptance, so close the pull request. After a better way found, will open another one to track.

**Pull request closed without merging**